File: python/P017_Letter_Combinations_of_a_Phone_Number/LetterCombinationsofaPhoneNumber.py
# coding=utf-8

from functools import reduce

class Solution(object):
    @staticmethod
    def letterCombinations(digits):
        if not digits or len(digits) == 0:
            return []
        phone = {'0': '',
               '1': '',
               '2': 'abc',
               '3': 'def',
               '4': 'ghi',
               '5': 'jkl',
               '6': 'mno',
               '7': 'pqrs',
               '8': 'tuv',
               '9': 'wxyz'}
        return reduce(lambda acc, digit: [x + y for x in acc for y in phone[digit]], digits, [''])

File: java/src/P017/LetterCombinationsofaPhoneNumber/LetterCombinationsOfAPhoneNumber.java
package P017.LetterCombinationsofaPhoneNumber;

/*
Given a digit string, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below.

Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
 */

import java.util.LinkedList;
import java.util.List;

public class LetterCombinationsOfAPhoneNumber {
    public static List<String> letterCombinations(String digits) {
        LinkedList<String> res = new LinkedList<>();
        if (digits == null || digits.length() == 0) return res;
        String[] mapping = new String[] {
                "0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
        res.add("");
        for (int i = 0; i < digits.length(); i++) {
            int x = Character.getNumericValue(digits.charAt(i));
            while (res.peek().length() == i) {
                String t = res.remove();
                for (char c : mapping[x].toCharArray()) {
                    res.add(t+c);
                }
            }
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println(letterCombinations("23"));
    }
}
